File: apps/runtime/src/listen.rs
//! Where the Runtime accepts HTTP and WebSocket requests — roadmap §4.4.
//!
//! Three transports, one router. A desktop install wants no TCP port at all:
//! the shell reaches the Runtime over a Unix socket (macOS / Linux) or a named
//! pipe (Windows), both of which the OS scopes to the current user and which
//! nothing on the network can reach. A browser install still needs TCP, so the
//! same router is served on a loopback listener whose port is normally handed
//! out by the kernel and published in `endpoints.json`.
//!
//! `--listen` takes one spec per occurrence:
//!
//! ```text
//! --listen tcp:127.0.0.1:0        kernel-assigned loopback port
//! --listen tcp:127.0.0.1:43120    a fixed port; already in use is an error
//! --listen unix:/abs/path.sock    Unix domain socket, 0600 in a 0700 directory
//! --listen pipe:armadra-xyz       \\.\pipe\armadra-xyz, first instance only
//! ```
//!
//! A spec for another platform's transport parses (so the same argument can be
//! tested everywhere) but fails at bind time with a message that says so,
//! rather than silently falling back to a port.

use std::{
    fmt,
    future::{Future, IntoFuture},
    net::SocketAddr,
    path::{Path, PathBuf},
    pin::Pin,
};

/// The Windows pipe namespace. A name is always rooted here; a spec may not
/// contain a separator of its own.
pub const PIPE_PREFIX: &str = r"\\.\pipe\";

/// One `--listen` argument.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ListenSpec {
    Tcp(SocketAddr),
    Unix(PathBuf),
    Pipe(String),
}

impl ListenSpec {
    /// Parses `tcp:ADDR:PORT`, `unix:PATH` or `pipe:NAME`. A bare `ADDR:PORT`
    /// is accepted as `tcp:` so `--listen 127.0.0.1:0` reads naturally.
    pub fn parse(spec: &str) -> Result<Self, String> {
        let spec = spec.trim();
        if let Some(rest) = spec.strip_prefix("tcp:") {
            return parse_tcp(rest);
        }
        if let Some(rest) = spec.strip_prefix("unix:") {
            // A leading `/`, not `Path::is_absolute`: the spec describes a Unix
            // socket path, and asking the *host* platform what "absolute" means
            // would make the same argument parse on Linux and be rejected on
            // Windows, where binding it is what has to fail, with a message
            // that says which transport is missing.
            if !rest.starts_with('/') {
                return Err(format!("--listen unix:PATH needs an absolute path: {rest}"));
            }
            return Ok(Self::Unix(PathBuf::from(rest)));
        }
        if let Some(rest) = spec.strip_prefix("pipe:") {
            return parse_pipe(rest);
        }
        parse_tcp(spec).map_err(|_| {
            format!("--listen expects tcp:ADDR:PORT, unix:PATH or pipe:NAME, not {spec:?}")
        })
    }

    /// The transport this spec asks for, independent of the address.
    pub fn kind(&self) -> &'static str {
        match self {
            Self::Tcp(_) => "tcp",
            Self::Unix(_) => "unix",
            Self::Pipe(_) => "pipe",
        }
    }
}

impl fmt::Display for ListenSpec {
    fn fmt(&self, out: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Tcp(address) => write!(out, "tcp:{address}"),
            Self::Unix(path) => write!(out, "unix:{}", path.display()),
            Self::Pipe(name) => write!(out, "pipe:{name}"),
        }
    }
}

fn parse_tcp(rest: &str) -> Result<ListenSpec, String> {
    rest.parse::<SocketAddr>()
        .map(ListenSpec::Tcp)
        .map_err(|_| format!("--listen tcp: needs an IP address and port, not {rest:?}"))
}

fn parse_pipe(rest: &str) -> Result<ListenSpec, String> {
    // The name is concatenated into a path. Anything that could climb out of
    // the pipe namespace, or that Windows would reject, is refused here so the
    // failure is the same message on every platform.
    let valid = !rest.is_empty()
        && rest.len() <= 200
        && rest
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.'));
    if !valid {
        return Err(format!(
            "--listen pipe:NAME accepts up to 200 characters of [A-Za-z0-9._-], not {rest:?}"
        ));
    }
    Ok(ListenSpec::Pipe(rest.to_owned()))
}

/// A listener that is already accepting. Created by [`bind`].
#[derive(Debug)]
pub enum BoundListener {
    Tcp(tokio::net::TcpListener),
    #[cfg(unix)]
    Unix(tokio::net::UnixListener, PathBuf),
    #[cfg(windows)]
    Pipe(windows_pipe::NamedPipeListener),
}

impl BoundListener {
    /// The TCP port the kernel actually gave us, for the endpoint files. `None`
    /// for every socket transport — those have no port by design.
    pub fn tcp_port(&self) -> Option<u16> {
        match self {
            Self::Tcp(listener) => listener.local_addr().ok().map(|address| address.port()),
            #[cfg(unix)]
            Self::Unix(..) => None,
            #[cfg(windows)]
            Self::Pipe(_) => None,
        }
    }

    /// The address this listener ended up on, in `--listen` spelling.
    pub fn resolved(&self) -> ListenSpec {
        match self {
            Self::Tcp(listener) => ListenSpec::Tcp(
                listener
                    .local_addr()
                    .unwrap_or_else(|_| SocketAddr::from(([0, 0, 0, 0], 0))),
            ),
            #[cfg(unix)]
            Self::Unix(_, path) => ListenSpec::Unix(path.clone()),
            #[cfg(windows)]
            Self::Pipe(listener) => ListenSpec::Pipe(listener.name().to_owned()),
        }
    }
}

/// Binds one spec. A TCP address that is already taken is an error and stays
/// one: roadmap §4.4 forbids quietly moving to the next port, because a client
/// that was told to use this address would then talk to nothing.
pub async fn bind(spec: &ListenSpec) -> anyhow::Result<BoundListener> {
    match spec {
        ListenSpec::Tcp(address) => match tokio::net::TcpListener::bind(address).await {
            Ok(listener) => Ok(BoundListener::Tcp(listener)),
            Err(error) if error.kind() == std::io::ErrorKind::AddrInUse => Err(anyhow::anyhow!(
                "{address} is already in use; stop the process holding it or pass \
                 --listen tcp:127.0.0.1:0 to let the kernel choose a port"
            )),
            Err(error) => {
                Err(anyhow::Error::new(error).context(format!("could not listen on {address}")))
            }
        },
        ListenSpec::Unix(path) => bind_unix(path),
        ListenSpec::Pipe(name) => bind_pipe(name),
    }
}

#[cfg(unix)]
fn bind_unix(path: &Path) -> anyhow::Result<BoundListener> {
    use anyhow::Context as _;

    if let Some(directory) = path.parent() {
        std::fs::create_dir_all(directory)
            .with_context(|| format!("could not create {}", directory.display()))?;
        crate::paths::harden_directory(directory);
    }
    // A socket file left by a Runtime that was killed would refuse the bind.
    // Anything actually listening is proven dead first: a connect that is
    // refused means no accepting process is behind this path.
    if let Ok(metadata) = std::fs::symlink_metadata(path) {
        use std::os::unix::fs::FileTypeExt as _;
        if !metadata.file_type().is_socket() {
            anyhow::bail!(
                "{} exists and is not a socket; refusing to replace it",
                path.display()
            );
        }
        match std::os::unix::net::UnixStream::connect(path) {
            Ok(stream) => {
                drop(stream);
                anyhow::bail!("another Runtime is already listening on {}", path.display());
            }
            Err(error) if error.kind() == std::io::ErrorKind::ConnectionRefused => {
                std::fs::remove_file(path).with_context(|| {
                    format!("could not remove the stale socket {}", path.display())
                })?;
            }
            Err(error) => {
                return Err(anyhow::Error::new(error).context(format!(
                    "cannot tell whether {} is stale; leaving it alone",
                    path.display()
                )));
            }
        }
    }
    let listener = tokio::net::UnixListener::bind(path)
        .with_context(|| format!("could not listen on {}", path.display()))?;
    // The socket is a full grant of the Runtime API; it is nobody else's.
    crate::paths::harden_file(path);
    Ok(BoundListener::Unix(listener, path.to_owned()))
}

#[cfg(not(unix))]
fn bind_unix(path: &Path) -> anyhow::Result<BoundListener> {
    anyhow::bail!(
        "--listen unix:{} needs a Unix-like platform; use --listen pipe:NAME here",
        path.display()
    )
}

#[cfg(windows)]
fn bind_pipe(name: &str) -> anyhow::Result<BoundListener> {
    windows_pipe::NamedPipeListener::bind(name).map(BoundListener::Pipe)
}

#[cfg(not(windows))]
fn bind_pipe(name: &str) -> anyhow::Result<BoundListener> {
    anyhow::bail!("--listen pipe:{name} needs Windows; use --listen unix:PATH here")
}

/// Serves `router` on `listener` until `shutdown` resolves.
///
/// Boxed because the three transports produce three unrelated future types and
/// `main` drives them all through one `try_join_all`.
pub fn serve(
    listener: BoundListener,
    router: axum::Router,
    shutdown: impl Future<Output = ()> + Send + 'static,
) -> Pin<Box<dyn Future<Output = std::io::Result<()>> + Send>> {
    match listener {
        BoundListener::Tcp(listener) => Box::pin(
            axum::serve(listener, router)
                .with_graceful_shutdown(shutdown)
                .into_future(),
        ),
        #[cfg(unix)]
        BoundListener::Unix(listener, _) => Box::pin(
            axum::serve(listener, router)
                .with_graceful_shutdown(shutdown)
                .into_future(),
        ),
        #[cfg(windows)]
        BoundListener::Pipe(listener) => Box::pin(
            axum::serve(listener, router)
                .with_graceful_shutdown(shutdown)
                .into_future(),
        ),
    }
}

/// Removes a Unix socket this process created. A path that is no longer ours —
/// replaced by a newer Runtime — is left alone.
pub fn release(listener_path: &Path) {
    #[cfg(unix)]
    {
        use std::os::unix::fs::FileTypeExt as _;
        if std::fs::symlink_metadata(listener_path)
            .is_ok_and(|metadata| metadata.file_type().is_socket())
        {
            let _ = std::fs::remove_file(listener_path);
        }
    }
    #[cfg(not(unix))]
    let _ = listener_path;
}

#[cfg(windows)]
pub mod windows_pipe {
    //! A named pipe that `axum::serve` can accept on.
    //!
    //! `first_pipe_instance` is what makes the name ours: a second process
    //! that tries to create the same pipe fails instead of silently receiving
    //! half of our clients. `reject_remote_clients` keeps the pipe off the
    //! network redirector. The default DACL grants the creating user, SYSTEM
    //! and administrators — the same reach a 0600 file in the user's data
    //! directory has.

    use tokio::net::windows::named_pipe::{NamedPipeServer, ServerOptions};

    pub struct NamedPipeListener {
        name: String,
        address: String,
        /// The instance the next `accept` will wait on. Always created before
        /// the previous one is handed out, so a client never finds the name
        /// momentarily absent.
        pending: Option<NamedPipeServer>,
    }

    // `NamedPipeServer` has no `Debug`, and the handle would say nothing
    // anyway; the pipe name is the part a log line needs.
    impl std::fmt::Debug for NamedPipeListener {
        fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            formatter
                .debug_struct("NamedPipeListener")
                .field("address", &self.address)
                .finish_non_exhaustive()
        }
    }

    impl NamedPipeListener {
        pub fn bind(name: &str) -> anyhow::Result<Self> {
            let address = format!("{}{name}", super::PIPE_PREFIX);
            let pending = ServerOptions::new()
                .first_pipe_instance(true)
                .reject_remote_clients(true)
                .create(&address)
                .map_err(|error| {
                    anyhow::Error::new(error)
                        .context(format!("could not create the named pipe {address}"))
                })?;
            Ok(Self {
                name: name.to_owned(),
                address,
                pending: Some(pending),
            })
        }

        pub fn name(&self) -> &str {
            &self.name
        }

        fn next_instance(&self) -> std::io::Result<NamedPipeServer> {
            ServerOptions::new()
                .reject_remote_clients(true)
                .create(&self.address)
        }
    }

    impl axum::serve::Listener for NamedPipeListener {
        type Io = NamedPipeServer;
        type Addr = String;

        async fn accept(&mut self) -> (Self::Io, Self::Addr) {
            loop {
                let server = match self.pending.take() {
                    Some(server) => server,
                    None => match self.next_instance() {
                        Ok(server) => server,
                        Err(error) => {
                            tracing::warn!(%error, "could not create a pipe instance");
                            tokio::time::sleep(std::time::Duration::from_millis(50)).await;
                            continue;
                        }
                    },
                };
                if let Err(error) = server.connect().await {
                    tracing::warn!(%error, "a pipe client went away before connecting");
                    continue;
                }
                self.pending = self.next_instance().ok();
                return (server, self.address.clone());
            }
        }

        fn local_addr(&self) -> std::io::Result<Self::Addr> {
            Ok(self.address.clone())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_transport_parses_on_every_platform() {
        assert_eq!(
            ListenSpec::parse("tcp:127.0.0.1:0"),
            Ok(ListenSpec::Tcp("127.0.0.1:0".parse().unwrap()))
        );
        // A bare address is the TCP shorthand.
        assert_eq!(
            ListenSpec::parse(" 127.0.0.1:9 "),
            Ok(ListenSpec::Tcp("127.0.0.1:9".parse().unwrap()))
        );
        assert_eq!(
            ListenSpec::parse("unix:/tmp/armadra/runtime.sock"),
            Ok(ListenSpec::Unix(PathBuf::from("/tmp/armadra/runtime.sock")))
        );
        assert_eq!(
            ListenSpec::parse("pipe:armadra-runtime.42"),
            Ok(ListenSpec::Pipe("armadra-runtime.42".into()))
        );
    }

    #[test]
    fn a_spec_round_trips_through_its_display_form() {
        for spec in [
            "tcp:127.0.0.1:43119",
            "unix:/tmp/armadra/runtime.sock",
            "pipe:armadra-runtime",
        ] {
            let parsed = ListenSpec::parse(spec).unwrap();
            assert_eq!(parsed.to_string(), spec);
            assert_eq!(ListenSpec::parse(&parsed.to_string()), Ok(parsed));
        }
    }

    #[test]
    fn addresses_that_could_escape_their_namespace_are_refused() {
        for spec in [
            "unix:relative/runtime.sock",
            "unix:",
            r"pipe:..\..\windows",
            "pipe:with space",
            "pipe:",
            "pipe:/etc/passwd",
            "tcp:localhost:1",
            "tcp:127.0.0.1",
            "https://127.0.0.1:1",
            "",
        ] {
            assert!(
                ListenSpec::parse(spec).is_err(),
                "{spec:?} should not have parsed"
            );
        }
        assert!(ListenSpec::parse(&format!("pipe:{}", "a".repeat(201))).is_err());
        assert!(ListenSpec::parse(&format!("pipe:{}", "a".repeat(200))).is_ok());
    }

    #[tokio::test]
    async fn a_taken_port_is_an_error_rather_than_a_different_port() {
        let held = tokio::net::TcpListener::bind("127.0.0.1:0").await.unwrap();
        let taken = held.local_addr().unwrap();
        let error = bind(&ListenSpec::Tcp(taken)).await.unwrap_err().to_string();
        assert!(error.contains("already in use"), "{error}");
        // And an unbound port still works, with the kernel's choice reported.
        let bound = bind(&ListenSpec::parse("tcp:127.0.0.1:0").unwrap())
            .await
            .unwrap();
        assert!(bound.tcp_port().is_some_and(|port| port != 0));
        assert_eq!(bound.resolved().kind(), "tcp");
    }

    #[cfg(unix)]
    #[tokio::test]
    async fn a_unix_socket_is_private_and_has_no_port() {
        use std::os::unix::fs::PermissionsExt;
        let directory = tempfile::tempdir().unwrap();
        let path = directory.path().join("nested").join("runtime.sock");
        let spec = ListenSpec::Unix(path.clone());
        let bound = bind(&spec).await.unwrap();
        assert_eq!(bound.tcp_port(), None);
        assert_eq!(bound.resolved(), spec);
        assert_eq!(
            std::fs::metadata(&path).unwrap().permissions().mode() & 0o777,
            0o600
        );
        assert_eq!(
            std::fs::metadata(path.parent().unwrap())
                .unwrap()
                .permissions()
                .mode()
                & 0o777,
            0o700
        );
        // A live socket is never stolen from the Runtime that owns it.
        let error = bind(&spec).await.unwrap_err().to_string();
        assert!(error.contains("already listening"), "{error}");
        drop(bound);
    }

    #[cfg(unix)]
    #[tokio::test]
    async fn a_stale_socket_is_replaced_but_a_regular_file_is_not() {
        let directory = tempfile::tempdir().unwrap();
        let path = directory.path().join("runtime.sock");
        {
            let listener = std::os::unix::net::UnixListener::bind(&path).unwrap();
            drop(listener);
        }
        // The inode survives the listener; binding it again must succeed.
        assert!(path.exists());
        let bound = bind(&ListenSpec::Unix(path.clone())).await.unwrap();
        drop(bound);
        release(&path);
        assert!(!path.exists());

        let occupied = directory.path().join("not-a-socket");
        std::fs::write(&occupied, b"important").unwrap();
        let error = bind(&ListenSpec::Unix(occupied.clone()))
            .await
            .unwrap_err()
            .to_string();
        assert!(error.contains("not a socket"), "{error}");
        assert_eq!(std::fs::read(&occupied).unwrap(), b"important");
    }

    #[cfg(not(windows))]
    #[tokio::test]
    async fn a_windows_pipe_spec_fails_here_instead_of_falling_back_to_a_port() {
        let error = bind(&ListenSpec::Pipe("armadra-runtime".into()))
            .await
            .unwrap_err()
            .to_string();
        assert!(error.contains("needs Windows"), "{error}");
    }
}

File: apps/desktop/src/core/listen.ts
import { connect } from "node:net";
import { existsSync, lstatSync, mkdirSync, unlinkSync } from "node:fs";
import type { Server } from "node:http";
import { dirname } from "node:path";
import { hardenDirectory, hardenFile } from "./paths";

/**
 * Where the core accepts HTTP and WebSocket requests.
 *
 * Three transports, one router. A desktop install wants no TCP port at all:
 * the shell reaches the core over a Unix socket (macOS / Linux) or a named
 * pipe (Windows), both of which the OS scopes to the current user and which
 * nothing on the network can reach. A browser install still needs TCP, so the
 * same router is served on a loopback listener whose port is normally handed
 * out by the kernel and published in `endpoints.json`.
 *
 * ```text
 * --listen tcp:127.0.0.1:0        kernel-assigned loopback port
 * --listen tcp:127.0.0.1:43120    a fixed port; already in use is an error
 * --listen unix:/abs/path.sock    Unix domain socket, 0600 in a 0700 directory
 * --listen pipe:armadra-xyz       \\.\pipe\armadra-xyz
 * ```
 *
 * A spec for another platform's transport parses (so the same argument can be
 * tested everywhere) and fails at bind time with a message that says so,
 * rather than silently falling back to a port. The parser is a byte-for-byte
 * port of `apps/runtime/src/listen.rs`, because `--listen` is an argument the
 * shell writes and either implementation reads.
 */

/** The Windows pipe namespace. A name is always rooted here. */
export const PIPE_PREFIX = "\\\\.\\pipe\\";

export type ListenSpec =
  | { readonly kind: "tcp"; readonly host: string; readonly port: number }
  | { readonly kind: "unix"; readonly path: string }
  | { readonly kind: "pipe"; readonly name: string };

export type ParseResult =
  | { readonly ok: true; readonly spec: ListenSpec }
  | { readonly ok: false; readonly reason: string };

/**
 * Parses `tcp:ADDR:PORT`, `unix:PATH` or `pipe:NAME`. A bare `ADDR:PORT` is
 * accepted as `tcp:` so `--listen 127.0.0.1:0` reads naturally.
 */
export function parseListenSpec(raw: string): ParseResult {
  const spec = raw.trim();
  if (spec.startsWith("tcp:")) return parseTcp(spec.slice(4));
  if (spec.startsWith("unix:")) {
    const path = spec.slice(5);
    // A leading `/`, not the host platform's idea of "absolute": the spec
    // describes a Unix socket path, and asking Windows what absolute means
    // would make the same argument parse on Linux and be rejected here, where
    // *binding* is what has to fail, with a message naming the transport.
    if (!path.startsWith("/")) {
      return {
        ok: false,
        reason: `--listen unix:PATH needs an absolute path: ${path}`,
      };
    }
    return { ok: true, spec: { kind: "unix", path } };
  }
  if (spec.startsWith("pipe:")) return parsePipe(spec.slice(5));
  const bare = parseTcp(spec);
  return bare.ok
    ? bare
    : {
        ok: false,
        reason: `--listen expects tcp:ADDR:PORT, unix:PATH or pipe:NAME, not ${JSON.stringify(spec)}`,
      };
}

function parseTcp(rest: string): ParseResult {
  const failure: ParseResult = {
    ok: false,
    reason: `--listen tcp: needs an IP address and port, not ${JSON.stringify(rest)}`,
  };
  const separator = rest.lastIndexOf(":");
  if (separator <= 0) return failure;
  const host = rest.slice(0, separator);
  const port = Number(rest.slice(separator + 1));
  if (
    !Number.isInteger(port) ||
    port < 0 ||
    port > 65535 ||
    rest.slice(separator + 1) === ""
  ) {
    return failure;
  }
  const literal = ipLiteral(host);
  if (literal === undefined) return failure;
  return { ok: true, spec: { kind: "tcp", host: literal, port } };
}

/**
 * The host half of a TCP spec, or `undefined` when it is not an IP literal.
 *
 * A hostname is refused the way Rust's `SocketAddr` refuses it: a spec that
 * resolved through DNS could name a machine on the network, and "loopback
 * only" would then depend on somebody's `/etc/hosts`.
 */
function ipLiteral(host: string): string | undefined {
  if (host.startsWith("[") && host.endsWith("]")) {
    const inner = host.slice(1, -1);
    return isIpv6(inner) ? host : undefined;
  }
  if (isIpv4(host)) return host;
  // A bare IPv6 address has its own colons, which the port split already ate.
  return undefined;
}

function isIpv4(host: string): boolean {
  const parts = host.split(".");
  return (
    parts.length === 4 &&
    parts.every(
      (part) =>
        /^\d{1,3}$/.test(part) && Number(part) >= 0 && Number(part) <= 255,
    )
  );
}

function isIpv6(host: string): boolean {
  // Enough to tell an address from a hostname: hex groups and at most one `::`.
  if (!/^[0-9A-Fa-f:.]+$/.test(host)) return false;
  return host.includes(":") && host.split("::").length <= 2;
}

function parsePipe(rest: string): ParseResult {
  // The name is concatenated into a path. Anything that could climb out of the
  // pipe namespace, or that Windows would reject, is refused here so the
  // failure reads the same on every platform.
  const valid =
    rest.length > 0 && rest.length <= 200 && /^[A-Za-z0-9._-]+$/.test(rest);
  if (!valid) {
    return {
      ok: false,
      reason: `--listen pipe:NAME accepts up to 200 characters of [A-Za-z0-9._-], not ${JSON.stringify(rest)}`,
    };
  }
  return { ok: true, spec: { kind: "pipe", name: rest } };
}

/** The `--listen` spelling of a spec; the inverse of `parseListenSpec`. */
export function formatListenSpec(spec: ListenSpec): string {
  switch (spec.kind) {
    case "tcp":
      return `tcp:${spec.host}:${spec.port}`;
    case "unix":
      return `unix:${spec.path}`;
    case "pipe":
      return `pipe:${spec.name}`;
  }
}

/** The authority a TCP spec's URLs are built from. */
export function tcpAuthority(spec: ListenSpec): string | undefined {
  return spec.kind === "tcp" ? `${spec.host}:${spec.port}` : undefined;
}

/**
 * Binds one spec on an already-created server.
 *
 * A TCP address that is already taken is an error and stays one: quietly
 * moving to the next port would leave a client that was told to use this
 * address talking to nothing.
 */
export async function bind(
  server: Server,
  spec: ListenSpec,
): Promise<ListenSpec> {
  switch (spec.kind) {
    case "tcp":
      return bindTcp(server, spec.host, spec.port);
    case "unix":
      return bindUnix(server, spec.path);
    case "pipe":
      return bindPipe(server, spec.name);
  }
}

function listening(server: Server, start: () => void): Promise<void> {
  return new Promise((resolve, reject) => {
    const onError = (error: Error): void => {
      server.off("listening", onListening);
      reject(error);
    };
    const onListening = (): void => {
      server.off("error", onError);
      resolve();
    };
    server.once("error", onError);
    server.once("listening", onListening);
    start();
  });
}

async function bindTcp(
  server: Server,
  host: string,
  port: number,
): Promise<ListenSpec> {
  const address = `${host}:${port}`;
  try {
    await listening(server, () =>
      server.listen({ host: host.replace(/^\[|\]$/g, ""), port }),
    );
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "EADDRINUSE") {
      throw new Error(
        `${address} is already in use; stop the process holding it or pass ` +
          `--listen tcp:127.0.0.1:0 to let the kernel choose a port`,
      );
    }
    throw new Error(`could not listen on ${address}: ${describe(error)}`);
  }
  const bound = server.address();
  if (bound === null || typeof bound === "string") {
    throw new Error(`could not read the bound address of ${address}`);
  }
  return { kind: "tcp", host, port: bound.port };
}

async function bindUnix(server: Server, path: string): Promise<ListenSpec> {
  if (process.platform === "win32") {
    throw new Error(
      `--listen unix:${path} needs a Unix-like platform; use --listen pipe:NAME here`,
    );
  }
  const directory = dirname(path);
  mkdirSync(directory, { recursive: true });
  hardenDirectory(directory);
  await clearStaleSocket(path);
  await listening(server, () => server.listen(path));
  // The socket is a full grant of the core's API; it is nobody else's.
  hardenFile(path);
  return { kind: "unix", path };
}

/**
 * A socket file left by a core that was killed would refuse the bind. Anything
 * actually listening is proven dead first: a connection that is refused means
 * no accepting process is behind this path.
 */
async function clearStaleSocket(path: string): Promise<void> {
  if (!existsSync(path)) return;
  if (!lstatSync(path).isSocket()) {
    throw new Error(
      `${path} exists and is not a socket; refusing to replace it`,
    );
  }
  const held = await socketIsHeld(path);
  if (held === "held") {
    throw new Error(`another core is already listening on ${path}`);
  }
  if (held === "unknown") {
    throw new Error(`cannot tell whether ${path} is stale; leaving it alone`);
  }
  unlinkSync(path);
}

function socketIsHeld(path: string): Promise<"held" | "stale" | "unknown"> {
  return new Promise((resolve) => {
    const socket = connect(path);
    const finish = (answer: "held" | "stale" | "unknown"): void => {
      socket.destroy();
      resolve(answer);
    };
    socket.setTimeout(1_000, () => finish("unknown"));
    socket.on("connect", () => finish("held"));
    socket.on("error", (error: NodeJS.ErrnoException) =>
      finish(error.code === "ECONNREFUSED" ? "stale" : "unknown"),
    );
  });
}

async function bindPipe(server: Server, name: string): Promise<ListenSpec> {
  if (process.platform !== "win32") {
    throw new Error(
      `--listen pipe:${name} needs Windows; use --listen unix:PATH here`,
    );
  }
  // TODO(R6): `node:net` creates an ordinary pipe instance. The first-instance
  // flag, `reject_remote_clients` and the per-connection client SID check the
  // Rust listener applies have no Node equivalent and need the session-host's
  // native layer; until then a Windows core is development-only.
  await listening(server, () => server.listen(`${PIPE_PREFIX}${name}`));
  return { kind: "pipe", name };
}

/**
 * Removes a Unix socket this process created. A path that is no longer ours —
 * replaced by a newer core — is left alone.
 */
export function release(spec: ListenSpec): void {
  if (spec.kind !== "unix") return;
  try {
    if (lstatSync(spec.path).isSocket()) unlinkSync(spec.path);
  } catch {
    // Already gone, or never ours.
  }
}

function describe(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
